File: melosys-eessi-app/src/test/java/no/nav/melosys/eessi/service/sed/mapper/til_sed/SedMapperTest.java
package no.nav.melosys.eessi.service.sed.mapper.til_sed;

import java.io.IOException;
import java.net.URISyntaxException;
import java.util.List;

import no.nav.melosys.eessi.controller.dto.Adressetype;
import no.nav.melosys.eessi.controller.dto.SedDataDto;
import no.nav.melosys.eessi.models.sed.nav.Adresse;
import no.nav.melosys.eessi.models.sed.nav.Arbeidsland;
import no.nav.melosys.eessi.models.sed.nav.Statsborgerskap;
import no.nav.melosys.eessi.service.sed.SedDataStub;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

class SedMapperTest {
    private final SedMapper sedMapper =  new TempTestSedMapper(); // fjerns når vi koverterer til Kotlin

    private SedDataDto sedData;

    @BeforeEach
    void setUp() throws IOException, URISyntaxException {
        sedData = SedDataStub.getStub();
    }

    @Test
    void hentAdresser() {
        final List<Adresse> adresser = sedMapper.hentAdresser(sedData);

        assertThat(adresser).hasSize(3)
            .anyMatch(adresse -> Adressetype.BOSTEDSADRESSE.getAdressetypeRina().equals(adresse.getType()))
            .anyMatch(adresse -> Adressetype.KONTAKTADRESSE.getAdressetypeRina().equals(adresse.getType()))
            .anyMatch(adresse -> Adressetype.POSTADRESSE.getAdressetypeRina().equals(adresse.getType()));
    }

    @Test
    void hentArbeidsland() {
        final List<Arbeidsland> arbeidsland = sedMapper.hentArbeidsland(sedData);

        assertThat(arbeidsland).hasSize(1);

        arbeidsland.stream().forEach(arbLand -> {
            assertThat(arbLand.getArbeidssted()).hasSize(1);
        });
    }

    @Test
    void hentStatsborgerskap() {
        final List<Statsborgerskap> statsborgerskap = sedMapper.hentStatsborgerskap(sedData);
        assertThat(statsborgerskap).hasSize(2).containsExactly(
            new Statsborgerskap("NO"),
            new Statsborgerskap("SE")
        );
    }
}

File: melosys-eessi-app/src/test/java/no/nav/melosys/eessi/service/sed/mapper/til_sed/SedMapperTest.kt
package no.nav.melosys.eessi.service.sed.mapper.til_sed

import io.kotest.matchers.collections.shouldContainExactly
import io.kotest.matchers.collections.shouldHaveSize
import io.kotest.matchers.nulls.shouldNotBeNull
import io.kotest.matchers.shouldBe
import no.nav.melosys.eessi.controller.dto.Adressetype
import no.nav.melosys.eessi.controller.dto.SedDataDto
import no.nav.melosys.eessi.models.SedType
import no.nav.melosys.eessi.models.sed.SED
import no.nav.melosys.eessi.models.sed.nav.Statsborgerskap
import no.nav.melosys.eessi.service.sed.SedDataStub
import org.junit.jupiter.api.Test

class SedMapperTest {
    private val sedMapper: SedMapper = object : SedMapper {
        override fun getSedType(): SedType = SedType.A003
    }

    private val sedData: SedDataDto = SedDataStub.getStub()

    @Test
    fun hentAdresser() {
        val adresser = sedMapper.hentAdresser(sedData)

        adresser.shouldHaveSize(3).map { it.type }.shouldContainExactly(
            Adressetype.BOSTEDSADRESSE.adressetypeRina,
            Adressetype.KONTAKTADRESSE.adressetypeRina,
            Adressetype.POSTADRESSE.adressetypeRina
        )
    }

    @Test
    fun hentArbeidsland() {
        val arbeidsland = sedMapper.hentArbeidsland(sedData)

        arbeidsland.shouldHaveSize(1)
            .single().arbeidssted.shouldHaveSize(1)
    }

    @Test
    fun hentStatsborgerskap() {
        val statsborgerskap = sedMapper.hentStatsborgerskap(sedData)

        statsborgerskap.shouldHaveSize(2).shouldContainExactly(
            Statsborgerskap("NO"),
            Statsborgerskap("SE")
        )
    }

    @Test
    fun `map tilleggsnavn til bygning for alle steder adresse blir brukt`() {
        val sed: SED = sedMapper.mapTilSed(SedDataStub.getStub {
            this.bostedsadresse!!.tilleggsnavn = "bostedsadresse Tilleggsnavn"
            this.kontaktadresse!!.tilleggsnavn = "kontaktadresse Tilleggsnavn"
            this.oppholdsadresse!!.tilleggsnavn = "oppholdsadresse Tilleggsnavn"
            arbeidsland.forEach {
                it.arbeidssted.forEach { arbeidssted ->
                    arbeidssted.adresse.tilleggsnavn = "Arbeidsland Tilleggsnavn"
                }
            }
        })
        sed.nav.shouldNotBeNull().run {
            bruker.shouldNotBeNull()
                .adresse.shouldNotBeNull().shouldHaveSize(3)
                .map { it.bygning }
                .shouldContainExactly(
                    "bostedsadresse Tilleggsnavn",
                    "kontaktadresse Tilleggsnavn",
                    "oppholdsadresse Tilleggsnavn"
                )

            arbeidsland.shouldNotBeNull().single()
                .arbeidssted.single()
                .adresse.shouldNotBeNull()
                .bygning shouldBe "Arbeidsland Tilleggsnavn"
        }
    }

    @Test
    fun `tilleggsnavn til bygning skal bli en gyldig EESSIMediumString`() {
        val sed: SED = sedMapper.mapTilSed(SedDataStub.getStub {
            this.bostedsadresse!!.tilleggsnavn = " "
            this.kontaktadresse!!.tilleggsnavn = (1..200).joinToString("") { "*" }
            this.oppholdsadresse!!.tilleggsnavn = "oppholdsadresse Tilleggsnavn"
            arbeidsland.forEach {
                it.arbeidssted.forEach { arbeidssted ->
                    arbeidssted.adresse.tilleggsnavn = " "
                }
            }
        })
        sed.nav.shouldNotBeNull().run {
            bruker.shouldNotBeNull()
                .adresse.shouldNotBeNull().shouldHaveSize(3)
                .map { it.bygning }
                .shouldContainExactly(
                    null,
                    (1..155).joinToString("") { "*" },
                    "oppholdsadresse Tilleggsnavn"
                )

            arbeidsland.shouldNotBeNull().single()
                .arbeidssted.single()
                .adresse.shouldNotBeNull()
                .bygning shouldBe null
        }
    }
}
